conn.go

```diff
@@ -0,0 +1,460 @@
+	defer c.mu.Unlock()
+
+	if c.closed {
+		return ErrClosed
+	}
+
+	for _, call := range c.pending {
+		close(call.done)
+	}
+
+	if cause != nil && cause != io.EOF && cause != io.ErrUnexpectedEOF {
+		c.logger.Printf("jsonrpc2: protocol error: %v\n", cause)
+	}
+
+	close(c.disconnect)
+	c.closed = true
+	return c.stream.Close()
+}
+
+func (c *Conn) readMessages(ctx context.Context) {
+	var err error
+	for err == nil {
+		var m anyMessage
+		err = c.stream.ReadObject(&m)
+		if err != nil {
+			break
+		}
+
+		switch {
+		case m.request != nil:
+			for _, onRecv := range c.onRecv {
+				onRecv(m.request, nil)
+			}
+			c.h.Handle(ctx, c, m.request)
+
+		case m.response != nil:
+			resp := m.response
+			if resp != nil {
+				id := resp.ID
+				c.mu.Lock()
+				call := c.pending[id]
+				delete(c.pending, id)
+				c.mu.Unlock()
+
+				if call != nil {
+					call.response = resp
+				}
+
+				if len(c.onRecv) > 0 {
+					var req *Request
+					if call != nil {
+						req = call.request
+					}
+					for _, onRecv := range c.onRecv {
+						onRecv(req, resp)
+					}
+				}
+
+				switch {
+				case call == nil:
+					c.logger.Printf("jsonrpc2: ignoring response #%s with no corresponding request\n", id)
+
+				case resp.Error != nil:
+					call.done <- resp.Error
+					close(call.done)
+
+				default:
+					call.done <- nil
+					close(call.done)
+				}
+			}
+		}
+	}
+	c.close(err)
+}
+
+func (c *Conn) send(_ context.Context, m *anyMessage, wait bool) (cc *call, err error) {
+	c.sending.Lock()
+	defer c.sending.Unlock()
+
+	// m.request.ID could be changed, so we store a copy to correctly
+	// clean up pending
+	var id ID
+
+	c.mu.Lock()
+	if c.closed {
+		c.mu.Unlock()
+		return nil, ErrClosed
+	}
+
+	// Assign a default id if not set
+	if m.request != nil && wait {
+		cc = &call{request: m.request, seq: c.seq, done: make(chan error, 1)}
+
+		isIDUnset := len(m.request.ID.Str) == 0 && m.request.ID.Num == 0
+		if isIDUnset {
+			if m.request.ID.IsString {
+				m.request.ID.Str = strconv.FormatUint(c.seq, 10)
+			} else {
+				m.request.ID.Num = c.seq
+			}
+		}
+		c.seq++
+	}
+	c.mu.Unlock()
+
+	if len(c.onSend) > 0 {
+		var (
+			req  *Request
+			resp *Response
+		)
+		switch {
+		case m.request != nil:
+			req = m.request
+		case m.response != nil:
+			resp = m.response
+		}
+		for _, onSend := range c.onSend {
+			onSend(req, resp)
+		}
+	}
+
+	// Store requests so we can later associate them with incoming
+	// responses.
+	if m.request != nil && wait {
+		c.mu.Lock()
+		id = m.request.ID
+		c.pending[id] = cc
+		c.mu.Unlock()
+	}
+
+	// From here on, if we fail to send this, then we need to remove
+	// this from the pending map so we don't block on it or pile up
+	// pending entries for unsent messages.
+	defer func() {
+		if err != nil {
+			if cc != nil {
+				c.mu.Lock()
+				delete(c.pending, id)
+				c.mu.Unlock()
+			}
+		}
+	}()
+
+	if err := c.stream.WriteObject(m); err != nil {
+		return nil, err
+	}
+	return cc, nil
+}
+
```

Comment:
I included the types that are below this comment in this file because they are only used by `Conn`.